### PR TITLE
Server Vault Client that Renews Token

### PR DESCRIPTION
**Pull request by dadgar**
This PR introduces:

* a test vault server
* A VaultClient interface
* An implementation of that interface that currently just contacts Vault, retrieves information about its token and then renews it

For review: @vishalnayak @diptanu

### CODE REVIEW

File: nomad/vault.go [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+
+	close(v.shutdownCh)
+	v.renewalRunning = false
+	v.establishingConn = false
+}
+
+// ConnectionEstablished returns whether a connection to Vault has been
+// established.
+func (v *vaultClient) ConnectionEstablished() bool {
+	v.l.Lock()
```
Comment (vishalnayak) [inline nomad/vault.go]: This can be a read lock.

File: nomad/vault.go [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+	backoff := 0.0
+
+	for {
+		select {
+		case <-v.shutdownCh:
+			return
+		case <-authRenewTimer.C:
+			// Renew the token and determine the new expiration
+			err := v.renew()
+			currentExpiration := v.lastRenewed.Add(time.Duration(v.token.CreationTTL) * time.Second)
```
Comment (vishalnayak) [inline nomad/vault.go]: If the token used in Nomad is *always* a periodic token, then we don't really need the lastRenewed to be stored.
Just `CreationTTL/2` should be sufficient.

The reason is, for expiring tokens, the returned lease duration may vary based on its `max_ttl`.
But for periodic tokens, the returned `LeaseDuration` is almost always the same (expect in cases where `explicit_max_ttl` is set along with `period`).

Comment (dadgar) [inline nomad/vault.go]: Is that true since the token may already have a TTL less that the CreationTTL by the time Nomad starts renewing it

Comment (vishalnayak) [inline nomad/vault.go]: The TTL of the token will be half the CreationTTL by the time Nomad renews it. But the increment for the renewal will always be CreationTTL and the response should contain the same value (+/- a second or two considering network delays). That's it.

Comment (dadgar) [inline nomad/vault.go]: I see!

**Pull request merged**